src/util/mount.c

```diff
@@ -51,22 +84,77 @@ int singularity_mount(const char *source, const char *target,
         /* NFS root_squash option set uid 0 to nobody, force use of real user ID */
         setfsuid(fsuid);
     }
-    ret = mount(source, target, filesystemtype, mountflags, data);
+
+    ret = mount(source, dest, filesystemtype, mountflags, data);
+    mount_errno = errno;
+
+    close(target_fd);
+    free(realdest);
+
     if ( singularity_priv_userns_enabled() == 0 && seteuid(singularity_priv_getuid()) < 0 ) {
         singularity_message(ERROR, "Failed to drop privileges: %s\n", strerror(errno));
         ABORT(255);
     }
 
+    errno = mount_errno;
     return ret;
 }
 
+int check_proc_mount(char *mount, char *real_mountpoint) {
+    int retval = -1;
+
+    char *dup_mountpoint = strdup(real_mountpoint);
+    char *test_mountpoint = dup_mountpoint;
+
+    while ( ( retval < 0 ) && ( strcmp(test_mountpoint, "/") != 0 ) ) {
+        char *full_test_path = NULL;
+        char *tmp_test_path = joinpath(CONTAINER_FINALDIR, test_mountpoint);
+
+        if ( is_link(tmp_test_path) == 0 ) {
+            char *linktarget = realpath(tmp_test_path, NULL); // Flawfinder: ignore
+            if ( linktarget == NULL ) {
+                singularity_message(ERROR, "Could not identify the source of contained link: %s\n", test_mountpoint);
+                ABORT(255);
+            }
+            full_test_path = joinpath(CONTAINER_FINALDIR, linktarget);
+            singularity_message(DEBUG, "Parent directory is a link, resolved: %s->%s\n", tmp_test_path, full_test_path);
+            if ( strcmp(linktarget, "/") == 0 ) {
+                singularity_message(DEBUG, "Link is pointing to /, not allowed: %s\n", test_mountpoint);
+                retval = 1;
+            }
+            free(linktarget);
+        } else {
+            full_test_path = strdup(tmp_test_path);
+        }
+
+        if( retval < 0 )
+        {
+            // Check to see if mountpoint is already mounted
+            if ( strcmp(full_test_path, mount) == 0 ) {
+                singularity_message(DEBUG, "Mountpoint is already mounted: %s\n", test_mountpoint);
+                retval = 1;
+            } else {
+                test_mountpoint = dirname(test_mountpoint);
+            }
+        }
+
+        free(tmp_test_path);
+        free(full_test_path);
+    }
+
+    free(dup_mountpoint);
+
+    return(retval);
+}
+
 int check_mounted(char *mountpoint) {
     int retval = -1;
     FILE *mounts;
     char *line = (char *)malloc(MAX_LINE_LEN);
-    char *rootfs_dir = CONTAINER_FINALDIR;
     unsigned int mountpoint_len = strlength(mountpoint, PATH_MAX);
-    char *real_mountpoint;
+    char *real_mountpoint = strdup(mountpoint);
```

Comment:
I don't see this real_mountpoint getting freed anywhere.

Reply:
Nevermind, it was in the original file, it just doesn't show up in the changeset.